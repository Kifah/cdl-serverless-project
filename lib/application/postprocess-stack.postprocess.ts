import {SQSEvent} from "aws-lambda";
import {AppConfigClient, GetConfigurationCommand} from "@aws-sdk/client-appconfig"; // ES Modules import

const client = new AppConfigClient({region: process.env.DEPLOY_REGION});

async function handler(event: SQSEvent) {

    const input = {
        "Application": "issues-app",
        "ClientId": "example-id",
        "Configuration": "notification-information",
        "Environment": process.env.DEPLOY_ENV
    };

    const command = new GetConfigurationCommand(input);

    // This fires after the blob has been read/loaded.
    const response = await client.send(command);
    const obfuscated = response.Content;
    const appConfigConfiguration = new TextDecoder().decode(obfuscated);
    console.log('region of app config:');
    console.log(process.env.DEPLOY_REGION);
    console.log('configs from parameter store:');
    console.log(process.env.PARAM_SORE_CONFIG);

    console.log('notification will be sent to (appConfig config):');
    console.log(appConfigConfiguration);
    //console.log(JSON.stringify(event));
    //console.log(event.Records[0].body);
}

export {handler}




