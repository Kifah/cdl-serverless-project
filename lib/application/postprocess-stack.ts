import * as cdk from 'aws-cdk-lib';
import {Construct} from 'constructs';
import * as sqs from 'aws-cdk-lib/aws-sqs';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import {NodejsFunction} from 'aws-cdk-lib/aws-lambda-nodejs';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import * as lambdaEventSources from 'aws-cdk-lib/aws-lambda-event-sources';

export enum DeployEnv {
    'test',
    'prod',
}

interface PostprocessStackProps extends cdk.StackProps {
    backendTable: dynamodb.ITable,
    deployEnv: DeployEnv

}

export class PostprocessStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props: PostprocessStackProps) {
        super(scope, id, props);


        const queue = new sqs.Queue(this, 'CdkPlayQueue', {
            visibilityTimeout: cdk.Duration.seconds(300)
        });

        // a pre-processing lambda receives messages from dynamoDB
        const preProcessHandler = new NodejsFunction(this, "CdkPlayPreProcessHandler", {
            runtime: lambda.Runtime.NODEJS_18_X,
            entry: 'resources/postprocess/preprocess.ts',
            handler: "handler",
            tracing: lambda.Tracing.ACTIVE,
            environment: {
                SQS_URL: queue.queueUrl,
            }
        });

        const postProcessHandler = new NodejsFunction(this, "CdkPlayPostProcessHandler", {
            runtime: lambda.Runtime.NODEJS_18_X,
            tracing: lambda.Tracing.ACTIVE,
            entry: 'resources/postprocess/postprocess.ts',
            handler: "handler",

        });

        const policyStatement = new cdk.aws_iam.PolicyStatement();
        policyStatement.addActions('appconfig:*');
        policyStatement.addResources('arn:aws:appconfig:eu-central-1:832476498399:*');
        postProcessHandler.addToRolePolicy(policyStatement);


        const tableEventSource = new lambdaEventSources.DynamoEventSource(props.backendTable, {
            startingPosition: lambda.StartingPosition.LATEST,
            filters: [lambda.FilterCriteria.filter({eventName: lambda.FilterRule.isEqual('INSERT')})],
        });
        preProcessHandler.addEventSource(tableEventSource);
        props.backendTable.grantStreamRead(preProcessHandler);
        queue.grantSendMessages(preProcessHandler);
        queue.grantConsumeMessages(postProcessHandler);
        const sqsEventSource = new lambdaEventSources.SqsEventSource(queue);
        postProcessHandler.addEventSource(sqsEventSource);

    }
}
