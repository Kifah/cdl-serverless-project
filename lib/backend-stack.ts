import * as cdk from 'aws-cdk-lib';
import {Construct} from 'constructs';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import {LambdaIntegration, RestApi} from "aws-cdk-lib/aws-apigateway";
import {NodejsFunction} from "aws-cdk-lib/aws-lambda-nodejs";
import {Runtime} from "aws-cdk-lib/aws-lambda";
import * as lambda from "aws-cdk-lib/aws-lambda";


export class BackendStack extends cdk.Stack {

    public readonly backendTable: dynamodb.ITable;

    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        //add dynamoDB table and give it streaming and permissions needed to forward messages to pre-process lambda
        this.backendTable = new dynamodb.Table(this, 'CdkPlayBackendTable', {
            partitionKey: {name: 'id', type: dynamodb.AttributeType.STRING},
            billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
            removalPolicy: cdk.RemovalPolicy.DESTROY,
            stream: dynamodb.StreamViewType.NEW_IMAGE
        });


        const backendCarsLambda = new NodejsFunction(this, 'CdkPlayBackendLambda', {
            runtime: lambda.Runtime.NODEJS_18_X,
            tracing: lambda.Tracing.ACTIVE,
            entry: 'resources/backend/backend.ts',
            handler: "handler",
            environment: {
                TABLE_NAME:this.backendTable.tableName
            }
        });

        const lambdaIntegration = new LambdaIntegration(backendCarsLambda);


        const backendRestApi = new RestApi(this, 'CdkPlayBackendApi');
        const spacesResource = backendRestApi.root.addResource('cars');
        spacesResource.addMethod('GET', lambdaIntegration);
        spacesResource.addMethod('POST', lambdaIntegration);

        this.backendTable.grantReadWriteData(backendCarsLambda);


    }
}