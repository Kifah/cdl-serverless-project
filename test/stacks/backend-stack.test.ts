import * as cdk from 'aws-cdk-lib';
import {Template} from 'aws-cdk-lib/assertions';
import * as Backend from '../../lib/application/backend-stack';


const app = new cdk.App();
const backendStack = new Backend.BackendStack(app, 'BackendTestStack');
const template = Template.fromStack(backendStack);

test('Has DynamoDB', () => {

    template.hasResourceProperties('AWS::DynamoDB::Table', {
        BillingMode: "PAY_PER_REQUEST"
    });

});

test("Matches the snapshot", () => {
    expect(template.toJSON()).toMatchSnapshot();
});



