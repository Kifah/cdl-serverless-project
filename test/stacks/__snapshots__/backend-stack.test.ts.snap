// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches the snapshot 1`] = `
{
  "Outputs": {
    "ApiEndpoint": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "CdkPlayBackendApi59FE35B6",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "CdkPlayBackendApiDeploymentStageprod162F2E75",
            },
            "/",
          ],
        ],
      },
    },
    "CdkPlayBackendApiEndpointC3DD9D09": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "CdkPlayBackendApi59FE35B6",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "CdkPlayBackendApiDeploymentStageprod162F2E75",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CdkPlayBackendApi59FE35B6": {
      "Properties": {
        "Name": "CdkPlayBackendApi",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "CdkPlayBackendApiAccount06A623BC": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "CdkPlayBackendApi59FE35B6",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "CdkPlayBackendApiCloudWatchRole088E4A25",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "CdkPlayBackendApiCloudWatchRole088E4A25": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "CdkPlayBackendApiDeploymentB7E31EF9f586cbe6ea705f18a3b882fc16b1510f": {
      "DependsOn": [
        "CdkPlayBackendApiissuesGET05A3C53D",
        "CdkPlayBackendApiissuesPOSTC57BD8CD",
        "CdkPlayBackendApiissuesBCF51D8A",
      ],
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "CdkPlayBackendApi59FE35B6",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "CdkPlayBackendApiDeploymentStageprod162F2E75": {
      "DependsOn": [
        "CdkPlayBackendApiAccount06A623BC",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "CdkPlayBackendApiDeploymentB7E31EF9f586cbe6ea705f18a3b882fc16b1510f",
        },
        "RestApiId": {
          "Ref": "CdkPlayBackendApi59FE35B6",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "CdkPlayBackendApiissuesBCF51D8A": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "CdkPlayBackendApi59FE35B6",
            "RootResourceId",
          ],
        },
        "PathPart": "issues",
        "RestApiId": {
          "Ref": "CdkPlayBackendApi59FE35B6",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "CdkPlayBackendApiissuesGET05A3C53D": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "CdkPlayBackendHandler12FA0478",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "CdkPlayBackendApiissuesBCF51D8A",
        },
        "RestApiId": {
          "Ref": "CdkPlayBackendApi59FE35B6",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "CdkPlayBackendApiissuesGETApiPermissionBackendTestStackCdkPlayBackendApi88127A6CGETissues13F77835": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "CdkPlayBackendHandler12FA0478",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "CdkPlayBackendApi59FE35B6",
              },
              "/",
              {
                "Ref": "CdkPlayBackendApiDeploymentStageprod162F2E75",
              },
              "/GET/issues",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "CdkPlayBackendApiissuesGETApiPermissionTestBackendTestStackCdkPlayBackendApi88127A6CGETissues1F7BD3DA": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "CdkPlayBackendHandler12FA0478",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "CdkPlayBackendApi59FE35B6",
              },
              "/test-invoke-stage/GET/issues",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "CdkPlayBackendApiissuesPOSTApiPermissionBackendTestStackCdkPlayBackendApi88127A6CPOSTissues317B63AF": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "CdkPlayBackendHandler12FA0478",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "CdkPlayBackendApi59FE35B6",
              },
              "/",
              {
                "Ref": "CdkPlayBackendApiDeploymentStageprod162F2E75",
              },
              "/POST/issues",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "CdkPlayBackendApiissuesPOSTApiPermissionTestBackendTestStackCdkPlayBackendApi88127A6CPOSTissuesC1AA079A": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "CdkPlayBackendHandler12FA0478",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "CdkPlayBackendApi59FE35B6",
              },
              "/test-invoke-stage/POST/issues",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "CdkPlayBackendApiissuesPOSTC57BD8CD": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "CdkPlayBackendHandler12FA0478",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "CdkPlayBackendApiissuesBCF51D8A",
        },
        "RestApiId": {
          "Ref": "CdkPlayBackendApi59FE35B6",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "CdkPlayBackendHandler12FA0478": {
      "DependsOn": [
        "CdkPlayBackendHandlerServiceRoleDefaultPolicy679F2451",
        "CdkPlayBackendHandlerServiceRole0D0B66FD",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "1441c3f2432d50bf3e7096f305456ed0bdc1f31110bbbbcfe3ca75231acfe2e3.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "TABLE_NAME": {
              "Ref": "CdkPlayBackendTable5A795537",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "CdkPlayBackendHandlerServiceRole0D0B66FD",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "CdkPlayBackendHandlerServiceRole0D0B66FD": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CdkPlayBackendHandlerServiceRoleDefaultPolicy679F2451": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "CdkPlayBackendTable5A795537",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CdkPlayBackendHandlerServiceRoleDefaultPolicy679F2451",
        "Roles": [
          {
            "Ref": "CdkPlayBackendHandlerServiceRole0D0B66FD",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CdkPlayBackendTable5A795537": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "PointInTimeRecoverySpecification": {
          "PointInTimeRecoveryEnabled": true,
        },
        "StreamSpecification": {
          "StreamViewType": "NEW_IMAGE",
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
